lib/sles4sap/console_redirection.pm

```diff
@@ -43,33 +43,133 @@ sub handle_login_prompt {
     # look for either password prompt or command prompt to appear
-    my $serial_response = wait_serial(qr/Password:\s*$|:~/, timeout => 20);
+    my $serial_response = wait_serial(qr/Password:\s*$|:~/, timeout => 20, quiet => 1);
 
     die 'Neither password not command prompt appeared.' unless $serial_response;
     # Handle password prompt if it appears
     if (grep /Password:\s*$/, $serial_response) {
         type_password $pwd;
         send_key 'ret';
         # wait for command prompt to be ready
-        die 'Command prompt did not appear within timeout' unless wait_serial(qr/:~|#|>/, timeout => 20);
+        die 'Command prompt did not appear within timeout' unless wait_serial(qr/:~|#|>/, timeout => 20, quiet => 1);
     }
     set_serial_term_prompt();    # set correct serial prompt
 }
 
 =head2 redirection_init
 
-    redirection_init();
+    redirection_init( [, ssh_user=>$ssh_user, target_ip=>$target_ip, ssh_tunnel_port=>$ssh_tunnel_port]);
+
+B<ssh_user>: Target login user - default value is defined by OpenQA parameter REDIRECT_TARGET_USER
+
+B<target_ip>: Target host IP - default value is defined by OpenQA parameter REDIRECT_TARGET_IP
+
+B<ssh_tunnel_port>: Local port used for reverse ssh tunnel, default: 22022
+
+Does initial setup for console redirection which includes:
+
+=over
+
+=item * sets BASE_VM_ID value with current machine id. This helps identifying the base (not redirected) VM.
 
-Do preparation before redirecting console. Gets base VM id and initial setup.
-This is required to be done only once at the beginning of the whole test.
-If you have a multi machine setup, execute this on each worker VM.
+=item * SSH key exchange for reverse ssh connection.
 
+=item * remote port forwarding for reverse SSH session
+
+=item * remote port forwarding of incoming traffic from remote VM to OpenQA server resources (upload_logs, download from 'data')
+
+=back
 =cut
 
 sub redirection_init {
+    my (%args) = @_;
+    $args{target_ip} //= get_required_var('REDIRECT_TARGET_IP');
+    $args{ssh_user} //= get_required_var('REDIRECT_TARGET_USER');
+    $args{ssh_tunnel_port} //= '22022';
+
+    croak 'Package autossh is not installed' if script_run('rpm -qi autossh');
+
+    record_info('Redirection init', 'Preparing console redirection');
+
     # This should get worker VM id before any redirection happening
     # ID serves as identification of the 'base' VM to return to.
-    set_var('WORKER_VM_ID', script_output 'cat /etc/machine-id');
+    set_var('BASE_VM_ID', script_output 'cat /etc/machine-id');
+
+    # Prepare keyless access from remote host to worker VM
+    connect_target_to_serial(%args);
+    script_run("sudo rm $reverse_ssh_key_base_name*", quiet => 1);
+    # remove any existing entry in known_hosts file
+    script_run("ssh-keygen -R [localhost]:$args{ssh_tunnel_port} -f ~/.ssh/known_hosts", quiet => 1);
+    script_run("sudo ssh-keygen -R [localhost]:$args{ssh_tunnel_port} -f /root/.ssh/known_hosts", quiet => 1);
```

Comment:
Maybe in the future we can try to extend `create_ssh_key` in `lib/publiccloud/provider.pm` for this

Reply:
considering it is executed after `connect_target_to_serial(%args);` , on which VM this command takes places?

Reply:
Exactly. This creates separate ssh key on Cloud host (Destination VM) and transfers public key to the Worker VM (Base VM) so you can use keyless ssh login from destination VM to base VM.